### PR TITLE
Resource leak checker full

### PR DESCRIPTION
**Pull request by kelloggm**

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+  /**
+   * {@link #makeExceptionalStores(MethodInvocationNode, TransferInput)} requires a TransferInput,
+   * but the actual exceptional stores need to be modified in {@link #accumulate(Node,
+   * TransferResult, String...)}, which only has access to a TransferResult. So this variable is set
+   * to non-null in {@link #visitMethodInvocation(MethodInvocationNode, TransferInput)} before the
+   * call to super, which will call accumulate(); this field is then reset to null afterwards to
+   * prevent it from being used somewhere it shouldn't be.
+   */
+  private @Nullable Map<TypeMirror, CFStore> exceptionalStores;
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java:52]: @kelloggm should this logic around propagating called methods facts along exceptional edges be moved to the main Called Methods Checker?  I thought we agreed that in general, a method should be considered "called" even if the call results in an exception.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            type, rlTypeFactory.calledMethodsValueElement, String.class);
+    List<String> newList =
+        Stream.concat(Arrays.stream(methodNames), currentMethods.stream())
+            .collect(Collectors.toList());
+
+    AnnotationMirror newType = rlTypeFactory.createCalledMethods(newList.toArray(new String[0]));
+    return newType;
+  }
+
+  private void handleEnsuresCalledMethodVarArgs(
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java:195]: @kelloggm similarly I think this logic can be moved to the main Called Methods Checker

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+
+  /** The CreatesObligation.value field/element. */
+  private final ExecutableElement createsObligationValueElement =
+      TreeUtils.getMethod(CreatesObligation.class, "value", 0, processingEnv);
+
+  /**
+   * Bidirectional map to preserve temporary variables created for nodes with non-empty @MustCall
+   * annotation and the corresponding nodes.
+   */
+  private BiMap<LocalVariableNode, Tree> tempVarToNode = HashBiMap.create();
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakAnnotatedTypeFactory.java:76]: @kelloggm @Nargeshdb does this map and the corresponding accessor methods belong in the `MustCallAnnotatedTypeFactory`?  We moved logic for creating temp variables there, so I wonder if it'd be cleaner to put this there as well

File: checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java [hunk 4]
```diff
@@ -0,0 +1,1428 @@
...
+
+/**
+ * Checks that all methods in {@link org.checkerframework.checker.mustcall.qual.MustCall} object
+ * types are invoked before the corresponding objects become unreachable
+ */
+/* package-private */
+class MustCallConsistencyAnalyzer {
+
+  /** By default, should we transfer ownership to the caller when a variable is returned? */
+  static final boolean TRANSFER_OWNERSHIP_AT_RETURN = true;
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallConsistencyAnalyzer.java:82]: @Nargeshdb should we get rid of this flag?  My feeling is that we don't really check it consistently in all the places that we need to.  And at this point I am fairly confident that we do want to assume that return types are `@Owning` by default.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java (hunk 1 above)
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakTransfer.java:52]: https://github.com/typetools/checker-framework/pull/4683

**Pull request closed without merging**